index.js

```diff
@@ -1,17 +1,36 @@
 'use strict'
 
 const fp = require('fastify-plugin')
-const { performance } = require('perf_hooks')
+const { performance, PerformanceObserver } = require('perf_hooks')
 const summary = require('summary')
 
 const ONSEND = 'on-send-'
-const PREHANDLER = 'pre-handler-'
+const ONREQUEST = 'on-request-'
 const ROUTES = 'fastify-routes:'
 
 module.exports = fp(async function (fastify, opts) {
-  fastify.addHook('preHandler', function (request, reply, next) {
+  let observedEntries = {}
+  const obs = new PerformanceObserver((items) => {
+    const fetchedItems = items.getEntries()
+    for (let i = 0; i < fetchedItems.length; i++) {
+      const e = fetchedItems[i]
+      if (e.name.indexOf(ROUTES) === 0) {
+        const key = e.name.split(':')[1]
+        if (observedEntries[key]) {
+          observedEntries[key].push(e.duration)
+        } else {
+          observedEntries[key] = [e.duration]
+        }
+      }
+    }
+
+    performance.clearMarks()
+  })
+  obs.observe({ entryTypes: ['measure'] })
+
+  fastify.addHook('onRequest', function (request, reply, next) {
     const id = request.raw.id
-    performance.mark(PREHANDLER + id)
+    performance.mark(ONREQUEST + id)
```

Comment:
Can you also add the request method in the key?

Reply:
If there is a `reply.context.config.statsId` eg. 'group-stats-together' would we expect this to still be separated by method, or together?  Or should we maybe add an additional config option say `reply.context.config.mergeMethods` ?  I think separated by the method makes the most sense here, but I suppose there could be a situation where one might want to join them all together.

Reply:
I think an option `mergeMethods would be good`.